backend/src/schema/resolvers/notifications.js

```diff
@@ -1,14 +1,80 @@
-import { neo4jgraphql } from 'neo4j-graphql-js'
+const resourceTypes = ['Post', 'Comment']
+
+const transformReturnType = record => {
+  return {
+    ...record.get('notification').properties,
+    from: {
+      __typename: record.get('resource').labels.find(l => resourceTypes.includes(l)),
+      ...record.get('resource').properties,
+    },
+    to: {
+      ...record.get('user').properties,
+    },
+  }
+}
 
 export default {
   Query: {
-    Notification: (object, params, context, resolveInfo) => {
-      return neo4jgraphql(object, params, context, resolveInfo, false)
+    notifications: async (parent, args, context, resolveInfo) => {
+      const { user: currentUser } = context
+      const session = context.driver.session()
+      let notifications
+      let whereClause
+      let orderByClause
+      switch (args.read) {
+        case true:
+          whereClause = 'WHERE notification.read = TRUE'
+          break
+        case false:
+          whereClause = 'WHERE notification.read = FALSE'
+          break
+        default:
+          whereClause = ''
+      }
+      switch (args.orderBy) {
+        case 'createdAt_asc':
+          orderByClause = 'ORDER BY notification.createdAt ASC'
+          break
+        case 'createdAt_desc':
+          orderByClause = 'ORDER BY notification.createdAt DESC'
+          break
+        default:
+          orderByClause = ''
+      }
+
+      try {
+        const cypher = `
+        MATCH (resource)-[notification:NOTIFIED]->(user:User {id:$id})
+        ${whereClause}
+        RETURN resource, notification, user
+        ${orderByClause}
+        `
+        const result = await session.run(cypher, { id: currentUser.id })
+        notifications = await result.records.map(transformReturnType)
+      } finally {
+        session.close()
+      }
+      return notifications
     },
   },
   Mutation: {
-    UpdateNotification: (object, params, context, resolveInfo) => {
-      return neo4jgraphql(object, params, context, resolveInfo, false)
+    markAsRead: async (parent, args, context, resolveInfo) => {
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-30T08:15:39Z" title="Friday, August 30th 2019, 10:15:39 am +02:00">Aug 30, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,14 +1,80 @@
-import { neo4jgraphql } from 'neo4j-graphql-js'
+const resourceTypes = ['Post', 'Comment']
+
+const transformReturnType = record => {
+  return {
+    ...record.get('notification').properties,
+    from: {
+      __typename: record.get('resource').labels.find(l => resourceTypes.includes(l)),
+      ...record.get('resource').properties,
+    },
+    to: {
+      ...record.get('user').properties,
+    },
+  }
+}
 
 export default {
   Query: {
-    Notification: (object, params, context, resolveInfo) => {
-      return neo4jgraphql(object, params, context, resolveInfo, false)
+    notifications: async (parent, args, context, resolveInfo) => {
+      const { user } = context
+      const session = context.driver.session()
+      let notifications
+      let whereClause
+      let orderByClause
+      switch (args.read) {
+        case true:
+          whereClause = 'WHERE notification.read = TRUE'
+          break
+        case false:
+          whereClause = 'WHERE notification.read = FALSE'
+          break
+        default:
+          whereClause = ''
+      }
+      switch (args.orderBy) {
+        case 'createdAt_asc':
+          orderByClause = 'ORDER BY notification.createdAt ASC'
+          break
+        case 'createdAt_desc':
+          orderByClause = 'ORDER BY notification.createdAt DESC'
+          break
+        default:
+          orderByClause = ''
+      }
+
+      try {
+        const cypher = `
+        MATCH (resource)-[notification:NOTIFIED]->(user:User {id:$id})
+        ${whereClause}
+        RETURN resource, notification, user
+        ${orderByClause}
+        `
+        const result = await session.run(cypher, { id: user.id })
+        notifications = await result.records.map(transformReturnType)
+      } finally {
+        session.close()
+      }
+      return notifications
     },
   },
   Mutation: {
-    UpdateNotification: (object, params, context, resolveInfo) => {
-      return neo4jgraphql(object, params, context, resolveInfo, false)
+    markAsRead: async (parent, args, context, resolveInfo) => {
```


The naming `markAsRead` has to less information to me.
If I read it out of much context I don’t know what this is about.
May be `markNotificationAsRead`!?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-30T10:44:30Z" title="Friday, August 30th 2019, 12:44:30 pm +02:00">Aug 30, 2019</time>_

---


You would *not* mark a **Noification** as read but a **Post** or **Comment**.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T12:20:22Z" title="Friday, August 30th 2019, 2:20:22 pm +02:00">Aug 30, 2019</time>_

---


that's a bit confusing for me @roschaefer since you have a `notifications` array which has an attribute `read` that starts as `false`

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-30T12:21:35Z" title="Friday, August 30th 2019, 2:21:35 pm +02:00">Aug 30, 2019</time>_

---


and in the query :point_up: you have a `notifications` query that has a switch on `args.read`, which checks if the attributed is true or false with a default...